lib/sysfs_class.c

```diff
@@ -60,13 +60,30 @@ void sysfs_close_class(struct sysfs_class *cls)
 	}
 }
 
+/*
+ * pass this function to dlist_find_custom()
+ * so it can compare device names
+ *
+ * return 1 if pathnames are equal, else 0
+ */
 static int cdev_name_equal(void *a, void *b)
 {
+	size_t length_a, length_b;
+	char *str_a, *str_b;
+
 	if (!a || !b)
 		return 0;
 
-	if (strncmp((char *)a, ((struct sysfs_class_device *)b)->name,
-				strlen((char *)a)) == 0)
+	str_a = (char *)a;
+	str_b = ((struct sysfs_class_device *)b)->name;
+
+	length_a = strnlen(str_a, SYSFS_NAME_LEN+1);
+	length_b = strnlen(str_b, SYSFS_NAME_LEN+1);
+
+	if (length_a != length_b)
+		return 0;
+
+	if (strncmp(str_a, str_b, SYSFS_NAME_LEN+1) == 0)
```

Comment:
Length of the string can be less than `SYSFS_NAME_LEN`, the `n` in the `strncmp` should be `length_a` or `length_b`. Throughout the code, the maximum allowed length is `SYSFS_NAME_LEN` which include `\0`, why is `SYSFS_NAME_LEN+1` required here for comparison, what am I missing here?

Reply:
Using length_a or length_b for strncmp() is exactly what makes this buggy. See the commit message: it will fail if the system has "eth1" and "eth10" -- i.e. "eth1" will not be shown. One can easily reproduce this by creating a system with >10 NICs and then doing:

# systool -c net | grep eth1
  Class Device = "eth10"
  Class Device = "eth11"

Reply:
Sorry for not being clear, `length_a` and `length_b` are equal after the check `if (length_a != length_b)` but can be lesser than `SYSFS_NAME_LEN`. Using one of them will reflect the comparison better, I don't have any strong opinion on the `n` here.

Reply:
Yes, I was pretty sure that using either length_a or length_b would work, since we know both strings are exactly that length. Or ... is it length_a+1? I always have to figure it out, again. And I knew both strings were less than SYSFS_NAME_LEN+1. :) But I admit I was being lazy. I'll append a commit that changes SYSFS_NAME_LEN with length_a, if that's okay with you (vs pulling this request and submitting a new one).

Reply:
> Using length_a or length_b for strncmp() is exactly what makes this buggy. See the commit message: it will fail if the system has "eth1" and "eth10" -- i.e. "eth1" will not be shown. One can easily reproduce this by creating a system with >10 NICs and then doing:
> 
> ```
> # systool -c net | grep eth1
>   Class Device = "eth10"
>   Class Device = "eth11"
> ```

I think using either length_a or length_b is okay once we establish they are both the same. Since "eth1" and "eth10" have different lengths, they will never get this far.

Reply:
Yes, sorry, you're both right. I seem to have ignored the fact that they were being compared right in the lines above.